### PR TITLE
test: update integration tests to run using Gitlab 16

### PR DESCRIPTION
**Pull request by TimKnight-DWP**
- Updates tests to use Gitlab 16 image
- Updates test to handle changes to the underlying APIs
- Adds programmatic date times to Access Token tests

### CODE REVIEW

File: tests/functional/conftest.py
Comment (nejch) [inline tests/functional/conftest.py:221]: @TimKnight-DWP I think I introduced this sidekiq helper years ago thinking it's a bit smarter than simple sleep calls. But I wonder if we should just delete this and instead have simpler helpers that wait for a resource to be in a certain state. E.g. wait for branch to exist/not exist, wait for MR to have certain status etc.

I'm just not sure how much work it would take, but it might be simpler.

**Review by nejch — COMMENTED** [review]
Thanks a lot for working on this @TimKnight-DWP!

I think this is based on some initial work by @JohnVillalovos so maybe cherry-picking or adding a co-authored-by trailer to include him might be good here. :+1: 

Also I started something related to bulk imports in https://github.com/python-gitlab/python-gitlab/pull/2494, so if you think you're close to getting this done, just mark the import/export tests as xfail and we can tackle them separately as a follow-up, just to unblock this upgrade. Let me know if that works!

File: tests/functional/conftest.py
Comment (TimKnight-DWP) [reply tests/functional/conftest.py:221]: I'll have a look, yeah the ones it mainly seems to matter for now are Merging MRs and I'm having to put more specific logic

I have John's MR open in a tab to guide me when I start on the `api` tests 👍

**Comment by TimKnight-DWP** [comment]
@nejch - something has clearly changed in user_agent which was causing failures when we did snipper.user_agent_details()

I can't see exactly what value that has in our tests, I will add one as a separate test marked as skipped - as I can't see anything in the changelogs as to what changed

**Comment by nejch** [comment]
@TimKnight-DWP if it only affects an individual test feel free to also skip it and we deal with it separately. GitLab often has subtle breaking changes in the API that are not announced - hence also some of the failures here.

**Comment by TimKnight-DWP** [comment]
Have done, separated it out into a distinct test so the feature can be validated when there's chance to come bacl. I also see a fgew open MRs around which look related to some of the remaining API failures

**Comment by nejch** [comment]
@TimKnight-DWP also if we are religiously asserting on something that is maybe not so relevant, we could change the tests a bit to really only check what matters. Especially with deleting resources, maybe we just test something simpler if the async deletions are a pain.

**Comment by TimKnight-DWP** [comment]
Sounds like a good shout, deletions definitely seem to be one of the biggest remaining pains, and always need a bit of time to sleep. 

I'm currently putting some logging in to see if there's an obvious "Scheduled for deletion" flag we can use rather than asserting thing is not there (maybe an if there -> assert scheduled for deletion, else assert has been deleted)

**Comment by TimKnight-DWP** [comment]
I think it may also. be worth validating the response from Delete APIs such as: https://docs.gitlab.com/ee/api/users.html#user-deletion

If GL returns a 204, we don't need to check that the thing has been immediately deleted, the api has told us yes, if it takes a while thats Gitlab functionality rather than python-gitlab functionaity

**Comment by nejch** [comment]
> I think it may also. be worth validating the response from Delete APIs such as: https://docs.gitlab.com/ee/api/users.html#user-deletion
> 
> If GL returns a 204, we don't need to check that the thing has been immediately deleted, the api has told us yes, if it takes a while thats Gitlab functionality rather than python-gitlab functionaity.
> 
> If my hunch is correct now, I'll do that for these failing users tests and then take a look through and swap over

Ah that's a great idea @TimKnight-DWP. We shouldn't be testing GitLab behavior, just that it receives the right requests. :+1: we could probably do that with a lot of other tests too, now that I think of it.

**Comment by TimKnight-DWP** [comment]
@nejch I think with this last push that should be tests Green 🤞 

I'll go through this aft and update all delete tests at least to just verify the API responded 2xx (i.e. didn't throw an exception) and then squash my commits

Should be good to go then 👍

File: tests/functional/api/test_boards.py
Comment (nejch) [inline tests/functional/api/test_boards.py:15]: @TimKnight-DWP I wonder if we really even need to wrap and re-raise this. The traceback we get from exposing the error might be more helpful than just a failing assert, since the error would be unexpected. WDYT? Then we just simplify and call the delete method.

This would apply to most other deletes as well, and the diff here would be a lot smaller too and it'll be easier for me to review. Sorry for the back and forth :bow:

File: tests/functional/api/test_merge_requests.py
Comment (nejch) [inline tests/functional/api/test_merge_requests.py:189]: I think we can delete the logs now if they were mostly used for debugging - same for any other logging added inside the actual test cases.

```suggestion
```

File: tests/functional/api/test_boards.py
Comment (TimKnight-DWP) [reply tests/functional/api/test_boards.py:15]: @nejch 🤔 if we did that it'd come up as an "Error" rather than a "Fail" 

I'm happy to go through and do that

File: tests/functional/api/test_merge_requests.py
Comment (TimKnight-DWP) [reply tests/functional/api/test_merge_requests.py:189]: Will go through my changelist and reduce logging and any other unnecessary changes

**Comment by TimKnight-DWP** [comment]
@nejch - would you be able to give this another review please? I think it'll also unblock adding integration tests for the CI Job Token PR

**Comment by TimKnight-DWP** [comment]
Showing as failing because -0.04% tests 😿

**Comment by nejch** [comment]
> Showing as failing because -0.04% tests 😿

That's ok @TimKnight-DWP I can override that (and fix the coverage in the code that is the root cause :sweat_smile:). I was focusing on something else at the moment, sorry!

**Comment by TimKnight-DWP** [comment]
@JohnVillalovos @nejch @max-wittig @bufferoverflow - this one is good to go in and then will be able to look at getting changes to modify the `job_token_scope` via python gitlab in, as GL have now released the APIs in a main build 👍

File: tests/functional/api/test_deploy_tokens.py
Comment (JohnVillalovos) [inline tests/functional/api/test_deploy_tokens.py:1]: nit: I'm personally not a fan of this type of import.

I prefer: `import datetime` and then just use `datetime.date.today()`

File: tests/functional/cli/test_cli_resource_access_tokens.py
Comment (JohnVillalovos) [inline tests/functional/cli/test_cli_resource_access_tokens.py:1]: Same nit as above.

File: tests/functional/cli/test_cli_users.py
Comment (JohnVillalovos) [inline tests/functional/cli/test_cli_users.py:1]: Same nit as above

File: tests/functional/cli/test_cli_v4.py
Comment (JohnVillalovos) [inline tests/functional/cli/test_cli_v4.py:3]: Same nit as above

File: tests/functional/conftest.py
Comment (JohnVillalovos) [inline tests/functional/conftest.py:7]: Same nit as above

File: tests/functional/conftest.py
Comment (JohnVillalovos) [inline tests/functional/conftest.py]: nit: s/space/spare/

File: tests/functional/conftest.py
Comment (JohnVillalovos) [inline tests/functional/conftest.py]: I'll be honest this looks concerning to me. As I think this will increase our chances of things randomly breaking because the process we wanted to finish is not finished.

File: tests/functional/helpers.py
Comment (JohnVillalovos) [inline tests/functional/helpers.py:42]: nit: Any reason not to use `if isinstance(object, gitlab.v4.objects.User)` ?

File: tests/functional/helpers.py
Comment (JohnVillalovos) [inline tests/functional/helpers.py:52]: nit: Any reason not to use `if isinstance(object, gitlab.v4.objects.Project)` ?

File: tests/functional/helpers.py
Comment (JohnVillalovos) [inline tests/functional/helpers.py:60]: This link doesn't seem correct.

File: tests/functional/helpers.py
Comment (JohnVillalovos) [inline tests/functional/helpers.py:64]: nit: Any reason not to use `if isinstance(object, gitlab.v4.objects.Group)` ?

File: tests/functional/api/test_merge_requests.py
Comment (JohnVillalovos) [inline tests/functional/api/test_merge_requests.py:2]: Same nit as before.

**Review by JohnVillalovos — COMMENTED** [review]
Left some comments and some nits.

I'm not sure I like the changes to `wait_for_sidekiq` and also how it isn't called anymore in a lot of places and just replaced with `sleep`. I worry that this will make the tests more likely to have random failures.

Also seems like we no longer test that `delete` actually deletes what we want deleted, we are testing that we can call delete and no error is raised though.

But on the positive we now have tests working with Gitlab 16, which is great.

So overall I'm not sure. I would like to hear what @nejch thinks.

**Comment by TimKnight-DWP** [comment]
With regards to no longer calling "delete", gitlab have significantly changed how they handle deletes, it happens way more asynchronously, and thus is really hard to actually account for in the tests without making them even more brittle (in the case of some deletes it happens after 1 day regardless). So all we can do is assert we successfully called the delete, and we have to trust Gitlab will act on the DELETE Rest request, rather than also try and test their functionality

File: tests/functional/helpers.py
Comment (TimKnight-DWP) [reply tests/functional/helpers.py:52]: nope - my lack of embedded python knowledge when I first did the MR 👍

**Comment by TimKnight-DWP** [comment]
It's been a month, but if I recall `wait_for_sidekiq` did a lot of checking assuming one free worker means we're good, but with the delete changes, gitlab stores those deletes in a Q, which is represented by checking how much concurrency sidekiq has used vs total. But when doing so I discovered we never got close to the limits due to threading on the container, so the check became somewhat redundant. The worker always showed as "busy" because it was waiting to perform deletes, but it had spare concurrency to perform other tasks.

The sleeps could probably go away to be honest, they are mainly there to help out the stability, sometimes the container/worker are a bit slower than expected so can randomly fail a test

**Comment by TimKnight-DWP** [comment]
@JohnVillalovos @nejch - any additional thoughts on this PR?

I think it's important for us to get the tests updated to 16, but also that unlocks some later PRs to add additional, new functionality only present in 16+

**Comment by TimKnight-DWP** [comment]
Hi @nejch @JohnVillalovos @max-wittig @bufferoverflow - sorry to tag all of you and to nudge this along, but with Gitlab 17 now around the corner, I'd like to get this in so the testing suite isn't > 2 versions behind the release, and to unblock some of the JOB_TOKEN_SCOPE proposed PRs, as people may need that functionality when GitLab start to deprecate the existing Token logic in 17.x

**Comment by JohnVillalovos** [comment]
I'm traveling at the moment, so don't have time to do the review.

File: tests/functional/cli/test_cli_v4.py
Comment (nejch) [inline tests/functional/cli/test_cli_v4.py:246]: Just so we're in sync here: this doesn't check the current approved status at this point in time, but whatever the state was at the time you created the MR object when fetching the data from the server with `list()`.

But maybe that's expected, as I'm not entirely sure what the purpose of this `if` check is? Perhaps dealing with pre-existing MRs if there's some race conditions in the project tests.

File: tests/functional/conftest.py
Comment (nejch) [inline tests/functional/conftest.py:393]: This doesn't actually print anything out though, right? Feels a bit wrong to just leave it here, unless I missed something. I think we can remove it if it was used to debug during this PR. or do we want to assert here that we really have approval rules?

**Review by nejch — COMMENTED** [review]
Thanks again @TimKnight-DWP for all the work and patience. I have a few more last comments and then I'd say we can get this merged just to unblock thins, and improve on the hardcoded sleeps later.

If you could also reword the commit messages to only say `chore` / `test` as this is not really user-facing changes that'd be great, but if it's too much hassle we can just squash the commits here in the end.

**Comment by nejch** [comment]
> Left some comments and some nits.
> 
> I'm not sure I like the changes to `wait_for_sidekiq` and also how it isn't called anymore in a lot of places and just replaced with `sleep`. I worry that this will make the tests more likely to have random failures.
> 
> Also seems like we no longer test that `delete` actually deletes what we want deleted, we are testing that we can call delete and no error is raised though.
> 
> But on the positive we now have tests working with Gitlab 16, which is great.
> 
> So overall I'm not sure. I would like to hear what @nejch thinks.

@JohnVillalovos I'm also not sure how I feel about removing _all_ the deletes, but in the end it won't be very reliable with current GitLab versions if we keep them. We might lose some breaking changes in the API this way, but this would be an upstream issue.

As for `wait_for_sidekiq`, I also think that was maybe not very reliable. If we find a more generic way, we can also point these tests against external gitlab instances in the future.

So I'd say lets try and get this done finally :sweat_smile: we can still open follow-up issues to improve on it.

File: tests/functional/api/test_deploy_tokens.py
Comment (TimKnight-DWP) [reply tests/functional/api/test_deploy_tokens.py:19]: @nejch yeah and/or I'll happily squash down to one prior to merge 👍

**Comment by TimKnight-DWP** [comment]
I can reproduce that smoke test failure locally on `main` branch too, will attempt to determine what change has caused it

**Comment by TimKnight-DWP** [comment]
@nejch updated with your comments, but something odd seems to be happening with the github-actions, and sudden unexpected failures in the smoke tests

**Comment by nejch** [comment]
@TimKnight-DWP the new smoke test failures seem unrelated as they also appear in https://github.com/python-gitlab/python-gitlab/pull/2833, might be something new in the build dependencies that change how the package is built, will need to check.

**Comment by nejch** [comment]
> Thanks @TimKnight-DWP! The new smoke test failures seem unrelated as they also appear in #2833, might be something new in the build dependencies that change how the package is built, will need to check.

The culprit seems to be https://github.com/pypa/build/issues/770. I'll adapt the tests.

**Review by nejch — APPROVED** [review]
Thanks again @TimKnight-DWP and @JohnVillalovos - I'll keep the commits as I see you've amended the messages and it's a pretty big PR :+1: 

Let's see if we can improve the `sleep()` calls in a follow-up. And thank for your patience :sweat_smile:

**Pull request merged**
